### DISCUSSION TITLE
Events

### DISCUSSION THREAD

**Discussion by Roms1383** [Halen84/ScriptHookRDR2DotNet-V2#14] [Q&A] [answered]
Hi @Halen84 ! Lately I looked into [RDR3 Native DB](https://rdr3.nativedb.dotindustries.dev/natives?search=event) to gather informations about the events of the game, and I'd like your feedback to get a better understanding.

What I'm interest into are probably the [eventType](https://github.com/Halen84/RDR3-Native-Flags-And-Enums/tree/main/eEventType)s, and I'd be interested to know if they can be listened to. Ideally I'd like to intercept all the events of the game, more specifically the interactions between player and NPCs.

A second specificity that I noticed but do not quite understand are methods like [TASK_SET_BLOCKING_OF_NON_TEMPORARY_EVENTS](https://rdr3.nativedb.dotindustries.dev/natives/0x90D2156198831D69) and their usage: what are the `NON_TEMPORARY` events ?

Thanks for your time :)

**Comment by Halen84** [comment, marked as answer]
Yes, events can be listened to, but it's very annoying. Each event has it's own unique struct, which means you'll need to know the size and members of each struct. Here's an example with `EVENT_HITCH_ANIMAL` in C++
```cpp
// Event Struct
struct sHitchAnimalEventData
{
	alignas(8) Ped Human;
	alignas(8) Ped Animal;
};

// Main Game Loop (called every frame)
int numberOfEvents = SCRIPTS::GET_NUMBER_OF_EVENTS(0);
for (int i = 0; i < numberOfEvents; i++)
{
	if (SCRIPTS::GET_EVENT_AT_INDEX(0, i) == MISC::GET_HASH_KEY("EVENT_HITCH_ANIMAL"))
	{
		sHitchAnimalEventData data;
		if (SCRIPTS::GET_EVENT_DATA(0, i, (Any*)&data, 4))
		{
			PED::SET_PED_TO_RAGDOLL(data.Human, 5000, 5000, 0, false, false, "");
		}
	}
}
```

`SET_BLOCKING_OF_NON_TEMPORARY_EVENTS` is used for if you want a ped to ignore reacting to permanent events, e.g decision maker.

**Reply by Roms1383** [reply]
Awesome @Halen84 ! ❤️ 
So, if I understand correctly, it's not a "traditional" event handler that we register to / unregister from which then triggers a callback function, but more like polling `SCRIPTS::GET_NUMBER_OF_EVENTS(eventGroupID)` regularly and processing queue's events.
> also on a side note, where can I find the mem repr for `sHitchAnimalEventData` ? I've looked a couple of repo but couldn't find any mention of its C++ definition.

Otherwise, when it comes to `NON_TEMPORARY` event I searched for [decision](https://rdr3.nativedb.dotindustries.dev/natives?search=decision) from your suggestion but [SET_DECISION_MAKER](https://rdr3.nativedb.dotindustries.dev/natives/0x8AE2F981CDDB8FA4) apparently expects a second argument of type `Hash` which is not documented, so I'm wondering what this is :)

**Reply by Halen84** [reply]
>it's not a "traditional" event handler that we register to / unregister from which then triggers a callback function, but more like polling SCRIPTS::GET_NUMBER_OF_EVENTS(eventGroupID) regularly and processing queue's events.

Yea exactly.

>where can I find the mem repr for `sHitchAnimalEventData` ?

Not sure what you mean by this. I have defined it in the above code.

>Otherwise, when it comes to NON_TEMPORARY event I searched for [decision](https://rdr3.nativedb.dotindustries.dev/natives?search=decision) from your suggestion but [SET_DECISION_MAKER](https://rdr3.nativedb.dotindustries.dev/natives/0x8AE2F981CDDB8FA4) apparently expects a second argument of type Hash which is not documented, so I'm wondering what this is :)

You can find a list of all decision makers in `events.meta` under the tags called `<Name>`

**Reply by Roms1383** [reply]
> Not sure what you mean by this. I have defined it in the above code.

Ah I meant: is there an exhaustive list out there with the memory representation / struct definitions for all events, like the one you described ?

**Reply by Halen84** [reply]
Not that I know of. You might be in the dark on that one. Looking at the decompiled scripts is probably your best bet

**Reply by Roms1383** [reply]
Thanks a lot @Halen84 :)

**Reply by Roms1383** [reply]
Just a little follow-up, I found tons of different versions of decompiled scripts across different Github repos, but searching for `EVENT_HITCH_ANIMAL` in any of them doesn't return any result, so I'm wondering if you found it out during your own experiments (e.g. in IDA), or maybe in a particular repository ? I wonder if I'm searching in wrong or outdated repos @Halen84

**Reply by Halen84** [reply]
Using IDA would help if you know where to look. If that doesn't help then you're just going to have to do lots of trial and error

**Reply by Roms1383** [reply]
Fair enough :)
